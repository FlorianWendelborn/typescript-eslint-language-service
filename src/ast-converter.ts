import * as ts from "typescript";
import { ParserOptions } from "@typescript-eslint/parser";
import { SourceCode, AST, Scope } from "eslint";
import { Extra } from "@typescript-eslint/typescript-estree/dist/parser-options";
import { analyzeScope } from "@typescript-eslint/parser/dist/analyze-scope";
import { simpleTraverse } from "@typescript-eslint/parser/dist/simple-traverse";
import { visitorKeys } from "@typescript-eslint/parser/dist/visitor-keys";
import { ParseAndGenerateServicesResult, TSESTreeOptions } from "@typescript-eslint/typescript-estree";
import convert from "@typescript-eslint/typescript-estree/dist/ast-converter";

function validateBoolean(value: boolean | undefined, fallback: boolean = false): boolean {
  if (typeof value !== 'boolean') {
    return fallback;
  }
  return value;
}

function createExtra(code: string) {
  const base = {
    tokens: null,
    range: false,
    loc: false,
    comment: false,
    comments: [],
    strict: false,
    jsx: false,
    useJSXTextNode: false,
    log: console.log,
    projects: [],
    errorOnUnknownASTType: false,
    errorOnTypeScriptSyntacticAndSemanticIssues: false,
    code: "",
    tsconfigRootDir: process.cwd(),
    extraFileExtensions: [],
    preserveNodeMaps: undefined,
  } as Extra;
  return {
    ...base,
    code,
    tokens: [],
    loc: true,
    comment: true,
    comments: [],
  } as Extra;
}

function applyParserOptionsToExtra(extra: Extra, options: TSESTreeOptions) {
  /**
   * Track range information in the AST
   */
  extra.range = typeof options.range === "boolean" && options.range;
  extra.loc = typeof options.loc === "boolean" && options.loc;
  /**
   * Track tokens in the AST
   */
  if (typeof options.tokens === "boolean" && options.tokens) {
    extra.tokens = [];
  }
  /**
   * Track comments in the AST
   */
  if (typeof options.comment === "boolean" && options.comment) {
    extra.comment = true;
    extra.comments = [];
  }
  /**
   * Enable JSX - note the applicable file extension is still required
   */
  if (typeof options.jsx === "boolean" && options.jsx) {
    extra.jsx = true;
  }
  /**
   * The JSX AST changed the node type for string literals
   * inside a JSX Element from `Literal` to `JSXText`.
   *
   * When value is `true`, these nodes will be parsed as type `JSXText`.
   * When value is `false`, these nodes will be parsed as type `Literal`.
   */
  if (typeof options.useJSXTextNode === "boolean" && options.useJSXTextNode) {
    extra.useJSXTextNode = true;
  }
  /**
   * Allow the user to cause the parser to error if it encounters an unknown AST Node Type
   * (used in testing)
   */
  if (
    typeof options.errorOnUnknownASTType === "boolean" &&
    options.errorOnUnknownASTType
  ) {
    extra.errorOnUnknownASTType = true;
  }
  /**
   * Allow the user to override the function used for logging
   */
  if (typeof options.loggerFn === "function") {
    extra.log = options.loggerFn;
  } else if (options.loggerFn === false) {
    extra.log = Function.prototype;
  }

  if (typeof options.project === "string") {
    extra.projects = [options.project];
  } else if (
    Array.isArray(options.project) &&
    options.project.every(projectPath => typeof projectPath === "string")
  ) {
    extra.projects = options.project;
  }

  if (typeof options.tsconfigRootDir === "string") {
    extra.tsconfigRootDir = options.tsconfigRootDir;
  }

  if (
    Array.isArray(options.extraFileExtensions) &&
    options.extraFileExtensions.every(ext => typeof ext === "string")
  ) {
    extra.extraFileExtensions = options.extraFileExtensions;
  }
  /**
   * Allow the user to enable or disable the preservation of the AST node maps
   * during the conversion process.
   *
   * NOTE: For backwards compatibility we also preserve node maps in the case where `project` is set,
   * and `preserveNodeMaps` is not explicitly set to anything.
   */
  extra.preserveNodeMaps =
    typeof options.preserveNodeMaps === "boolean" && options.preserveNodeMaps;
  if (options.preserveNodeMaps === undefined && extra.projects.length > 0) {
    extra.preserveNodeMaps = true;
  }

  return extra;
}

export type AstConverterCreateOptions = {
  program: ts.Program,
};

export class AstConverter {
  public program: ts.Program;

  constructor({ program }: AstConverterCreateOptions) {
    this.program = program;
  }

  /**
   *
   * see also https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts#L346
   *
   */
  parseAndGenerateServices(src: ts.SourceFile, options: TSESTreeOptions) {
    const code = src.getFullText();
    let extra: Extra = createExtra(code);

    extra.code = code;

    /**
     * Apply the given parser options
     */
    if (typeof options !== "undefined") {
      extra = applyParserOptionsToExtra(extra, options);
      if (typeof options.errorOnTypeScriptSyntacticAndSemanticIssues === "boolean" && options.errorOnTypeScriptSyntacticAndSemanticIssues) {
        extra.errorOnTypeScriptSyntacticAndSemanticIssues = true;
      }
    }
    const { estree, astMaps } = convert(src, extra, true);

    /**
     * Return the converted AST and additional parser services
     */
    return {
      ast: estree,
      services: {
        program: this.program,
        esTreeNodeToTSNodeMap: astMaps ? astMaps.esTreeNodeToTSNodeMap : undefined,
        tsNodeToESTreeNodeMap: astMaps ? astMaps.tsNodeToESTreeNodeMap : undefined,
      },
    } as ParseAndGenerateServicesResult<any>;
  }

  /**
   * See also https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/parser/src/parser.ts
   */
  parseForESLint(src: ts.SourceFile, options?: ParserOptions | null) {
    try {
      
      if (!options || typeof options !== "object") {
        options = {};
      }

      if (options.sourceType !== "module" && options.sourceType !== "script") {
        options.sourceType = "script";
      }
      if (typeof options.ecmaFeatures !== "object") {
        options.ecmaFeatures = {};
      }

      const parserOptions: TSESTreeOptions = {};
      Object.assign(parserOptions, options, {
        useJSXTextNode: validateBoolean(options.useJSXTextNode, true),
        jsx: validateBoolean(options.ecmaFeatures.jsx),
      });

      if (typeof options.filePath === "string") {
        const tsx = options.filePath.endsWith(".tsx");
        if (tsx || options.filePath.endsWith(".ts")) {
          parserOptions.jsx = tsx;
        }
      }

      const { ast, services } = this.parseAndGenerateServices(src, parserOptions);

      simpleTraverse(ast, {
        enter(node) {
          switch (node.type) {
            // Function#body cannot be null in ESTree spec.
            case 'FunctionExpression':
              if (!node.body) {
              node.type = `TSEmptyBody${node.type}` as any;
            }
            break;
            // no default
          }
        },
      });

      const scopeManager = analyzeScope(ast, parserOptions);

      return {
        ast: ast as AST.Program,
        scopeManager: scopeManager as Scope.ScopeManager,
        services,
        visitorKeys,
      };
    } catch (error) {
      throw new Error();
    }
  }

  convertToESLintSourceCode(src: ts.SourceFile, options?: ParserOptions | null) {

    const code = src.getFullText();
    const { ast, scopeManager } = this.parseForESLint(src, options);

    return new SourceCode({
      text: code,
      ast,
      scopeManager,
    });
  }
}
