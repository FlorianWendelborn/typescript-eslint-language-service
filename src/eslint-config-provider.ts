import path from "path";
import * as ts from "typescript";
import { CascadingConfigArrayFactory } from "eslint/lib/cli-engine/cascading-config-array-factory";

export type ConfigProviderHost = {
  readonly readFile: (fileName: string, encoding: string) => string | undefined;
  readonly watchFile: (path: string, callback: ts.FileWatcherCallback, pollingInterval?: number) => ts.FileWatcher;
};

export type ESLintConfigProviderOptions = {
  host: ConfigProviderHost;
  directoriesToWatch: string[];
};

const ESLINTRC_SUFFIX_LIST = [
  ".eslintrc.js",
  ".eslintrc.yaml",
  ".eslintrc.yml",
  ".eslintrc.json",
  ".eslintrc",
  "package.json",
];

export class ESLintConfigProvider {

  private readonly host: ConfigProviderHost;
  private factory: CascadingConfigArrayFactory;

  public constructor({ host, directoriesToWatch }: ESLintConfigProviderOptions) {

    this.host = host;
    this.factory = new CascadingConfigArrayFactory();

    directoriesToWatch.forEach(directory => {
      ESLINTRC_SUFFIX_LIST.map(suffix => path.resolve(directory, suffix)).forEach(eslintrcFilepath => {
        this.host.watchFile(eslintrcFilepath, () => this.factory.clearCache(), 50);
      });
    });
  }

  public getConfigArrayForFile(fileName: string) {
    return this.factory.getConfigArrayForFile(fileName);
  }

  public getConfigForFile(fileName: string) {
    return this.factory.getConfigArrayForFile(fileName).extractConfig(fileName);
  }
}
