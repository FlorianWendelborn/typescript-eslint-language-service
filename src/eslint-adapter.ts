import * as ts from "typescript";
import { Linter } from "eslint";
import { AstConverter } from "./ast-converter";
import { InvalidParserError } from "./errors";
import { ESLintConfigProvider } from "./eslint-config-provider";

// TODO refactor global const
export const TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE = 30010;

export function translateESLintResult(result: Linter.LintMessage[], sourceFile: ts.SourceFile): ts.Diagnostic[] {
  return result.map(({ message, severity, ruleId, line, column, endLine, endColumn })=> {
    const messageText = `[${ruleId}] ${message}`;

    const category: ts.DiagnosticCategory = 
      severity === 2 ? ts.DiagnosticCategory.Error :
        severity === 1 ? ts.DiagnosticCategory.Warning :
          ts.DiagnosticCategory.Suggestion;

    /**
     * ESLint uses 1-started index. On the other hand, TypeScript 0-started index.
     * So we should minus from ESLint result's row/column to create TypeScript position.
     */
    const start = ts.getPositionOfLineAndCharacter(sourceFile, line - 1, column - 1);
    const end = endLine && endColumn ? ts.getPositionOfLineAndCharacter(sourceFile, endLine - 1, endColumn -1) : start;
    const length = Math.max(0, end - start);

    const diagnostic: ts.Diagnostic = {
      file: sourceFile,
      category,
      code: TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE,
      messageText,
      start,
      length,
    };

    return diagnostic;
  });
}

export type ESLintAdapterOptions = {
  logger: (msg: string) => void;
  getSourceFile: (fileName: string) => ts.SourceFile | undefined;
  converter: AstConverter;
  configProvider: ESLintConfigProvider;
};

export class ESLintAdapter {
  linter: Linter;
  logger: (msg: string) => void;
  converter: AstConverter;
  configProvider: ESLintConfigProvider;
  getSourceFile: (fileName: string) => ts.SourceFile | undefined;

  constructor({
    logger,
    converter,
    configProvider,
    getSourceFile,
  }: ESLintAdapterOptions) {
    this.linter = new Linter();
    this.logger = logger;
    this.converter = converter;
    this.configProvider = configProvider;
    this.getSourceFile = getSourceFile;
  }

  getSemanticDiagnostics(delegate: ts.LanguageService["getSemanticDiagnostics"], fileName: string): ReturnType<ts.LanguageService["getSemanticDiagnostics"]> {
    const original = delegate(fileName);
    try {
      const sourceFile = this.getSourceFile(fileName);
      if (!sourceFile) {
        return original;
      }
      const configArray = this.configProvider.getConfigArrayForFile(fileName);
      const configFileContent = configArray.extractConfig(fileName).toCompatibleObjectAsConfigFileContent();
      if (!configFileContent.parser || !/@typescript-eslint\/parser/.test(configFileContent.parser)) {
        throw new InvalidParserError();
      }
      const parserOptions = configFileContent.parserOptions ? configFileContent.parserOptions : { };
      const sourceCode = this.converter.convertToESLintSourceCode(sourceFile, parserOptions);

      // See https://github.com/eslint/eslint/blob/v6.1.0/lib/linter/linter.js#L1130
      const eslintResult = this.linter.verify(sourceCode, configArray as any, { filename: fileName });

      return [...original, ...translateESLintResult(eslintResult, sourceFile)];
    } catch (error) {
      this.logger(error.message ? error.message : "unknow error");
      return original;
    }
  }
}
