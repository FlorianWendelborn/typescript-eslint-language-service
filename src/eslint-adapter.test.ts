import path from "path";
import ts from "typescript";
import { AstConverter } from "./ast-converter";
import { ESLintAdapter } from "./eslint-adapter";
import { ConfigProvider } from "./eslint-config-provider";
import { ConfigArray } from "eslint/lib/cli-engine/config-array/config-array";

class TestingConfigProvider implements ConfigProvider {
  public conf: any = {
    parser: {
      error: null,
      id: "@typescript-eslint/parser",
      filePath: path.resolve(__dirname, "../node_modules/@typescript-eslint/parser/dist/parser.js"),
    },
  };
  public getConfigArrayForFile(): ConfigArray {
    return new ConfigArray(this.conf);
  }
}

function filterDiag(diagnostics: ts.Diagnostic[]) {
  // eslint-disable-next-line
  return diagnostics.map(({ file, ...rest }: ts.Diagnostic) => ({ ...rest }));
}

describe("ESLintAdapter", () => {
  describe("#getSemanticDiagnostics", () => {
    it("shuld return ESLint verification result as TypeScript diagnostic format", () => {
      const configProvider = new TestingConfigProvider();
      configProvider.conf.rules = {
        semi: 2,
      };
      const adapter = new ESLintAdapter({
        converter: new AstConverter(),
        getSourceFile: () => ts.createSourceFile("main.ts", "const x = 1", ts.ScriptTarget.ESNext, true),
        configProvider,
        logger: () => { },
      });
      const diagnostics = filterDiag(adapter.getSemanticDiagnostics(() => [], "main.ts"));
      expect(diagnostics).toMatchSnapshot();
    });
  });
});
