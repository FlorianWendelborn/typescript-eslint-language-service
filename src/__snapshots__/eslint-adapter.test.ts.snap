// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESLintAdapter #getCodeFixesAtPosition shuld only delegate when input errorCodes does not include TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE 1`] = `Array []`;

exports[`ESLintAdapter #getCodeFixesAtPosition shuld only delegate when input range isn't intersected ESLint reported results 1`] = `Array []`;

exports[`ESLintAdapter #getCodeFixesAtPosition shuld return ESLint sourceCodeFixer result as TypeScript codeFixAction format 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "fileName": "main.ts",
        "isNewFile": false,
        "textChanges": Array [
          Object {
            "newText": "\\"use strict\\"",
            "span": Object {
              "length": 12,
              "start": 11,
            },
          },
        ],
      },
    ],
    "description": "Fix: Strings must use doublequote.",
    "fixId": "quotes",
    "fixName": "quotes",
  },
]
`;

exports[`ESLintAdapter #getSemanticDiagnostics shuld return ESLint verification result as TypeScript diagnostic format 1`] = `
Array [
  Object {
    "category": 1,
    "code": 30010,
    "length": 0,
    "messageText": "[semi] Missing semicolon.",
    "start": 11,
  },
]
`;
